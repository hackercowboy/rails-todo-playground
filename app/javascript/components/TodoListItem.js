import React from 'react';
import PropTypes from 'prop-types';

const statusButtonClassMapping = {
  pending: 'update-button',
  done: 'delete-button',
};

const statusButtonTextMapping = {
  pending: 'Done',
  done: 'Delete',
};

function TodoListItem({
  id, status, text, onDone, onDelete,
}) {
  const handleOnClick = () => {
    if (status === 'pending') {
      onDone(id);
    }

    if (status === 'done') {
      onDelete(id);
    }
  };

  return (
    <li className="todo-list-item">
      <span className={`todo-list-item-text ${status}`}>{text}</span>
      <span className={`todo-list-item-action ${status}`}>
        <button className={statusButtonClassMapping[status]} onClick={handleOnClick}>
          {statusButtonTextMapping[status]}
        </button>
      </span>
    </li>);
}

TodoListItem.propTypes = {
  id: PropTypes.number.isRequired,
  status: PropTypes.oneOf(['pending', 'done']).isRequired,
  text: PropTypes.string.isRequired,
  onDelete: PropTypes.func.isRequired,
  onDone: PropTypes.func.isRequired,
};

export default TodoListItem;
