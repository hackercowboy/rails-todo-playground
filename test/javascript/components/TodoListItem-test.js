import React from 'react';
import { shallow } from 'enzyme';

import TodoListItem from '../../../app/javascript/components/TodoListItem';

describe('<TodoListItem/>', () => {
  it('should render', () => {
    const wrapper = shallow(<TodoListItem id={32} status="pending" text="Test" onDone={jest.fn()} onDelete={jest.fn()} />);
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should call onDone', () => {
    const handleOnDone = jest.fn();
    const wrapper = shallow(<TodoListItem id={32} status="pending" text="Test" onDone={handleOnDone} onDelete={jest.fn()} />);
    wrapper.find('button').props().onClick();
    expect(handleOnDone).toHaveBeenCalledWith(32);
  });

  it('should call onDelete', () => {
    const handleOnDelete = jest.fn();
    const wrapper = shallow(<TodoListItem id={32} status="done" text="Test" onDone={jest.fn()} onDelete={handleOnDelete} />);
    wrapper.find('button').props().onClick();
    expect(handleOnDelete).toHaveBeenCalledWith(32);
  });
});
