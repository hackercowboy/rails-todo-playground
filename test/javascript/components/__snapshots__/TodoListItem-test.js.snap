// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoListItem/> should render 1`] = `
"<li className=\\"todo-list-item\\">
  <span className=\\"todo-list-item-text pending\\">
    $
    Test
  </span>
  <span className=\\"todo-list-item-action pending\\">
    <button className=\\"update-button\\" onClick={[Function: handleOnClick]}>
      $
      Done
    </button>
  </span>
</li>"
`;
